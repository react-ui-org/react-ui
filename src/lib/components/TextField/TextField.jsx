import PropTypes from 'prop-types';
import React, { useContext } from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootSizeClassName } from '../_helpers/getRootSizeClassName';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { resolveContextOrProp } from '../_helpers/resolveContextOrProp';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import withForwardedRef from '../withForwardedRef';
import styles from './TextField.scss';

const SMALL_INPUT_SIZE = 10;

export const TextField = ({
  disabled,
  forwardedRef,
  fullWidth,
  helpText,
  id,
  inputSize,
  isLabelVisible,
  label,
  layout,
  placeholder,
  required,
  size,
  type,
  validationState,
  validationText,
  value,
  variant,
  ...restProps
}) => {
  const context = useContext(FormLayoutContext);
  const hasSmallInput = (inputSize !== null) && (inputSize <= SMALL_INPUT_SIZE);

  return (
    <label
      className={classNames(
        styles.root,
        fullWidth && styles.isRootFullWidth,
        hasSmallInput && styles.hasRootSmallInput,
        inputSize && styles.hasRootCustomInputSize,
        context && styles.isRootInFormLayout,
        resolveContextOrProp(context && context.layout, layout) === 'horizontal'
          ? styles.rootLayoutHorizontal
          : styles.rootLayoutVertical,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootSizeClassName(size, styles),
        getRootValidationStateClassName(validationState, styles),
        variant === 'filled' ? styles.rootVariantFilled : styles.rootVariantOutline,
      )}
      htmlFor={id}
      id={id && `${id}__label`}
      {...(inputSize ? { style: { '--rui-custom-input-size': inputSize } } : {})}
    >
      <div
        className={classNames(
          styles.label,
          !isLabelVisible && styles.isLabelHidden,
        )}
        id={id && `${id}__labelText`}
      >
        {label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <input
            {...transferProps(restProps)}
            className={styles.input}
            disabled={disabled}
            id={id}
            placeholder={placeholder}
            ref={forwardedRef}
            required={required}
            size={type !== 'number' ? inputSize : null}
            type={type}
            value={value}
          />
          {variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
        {helpText && (
          <div
            className={styles.helpText}
            id={id && `${id}__helpText`}
          >
            {helpText}
          </div>
        )}
        {validationText && (
          <div
            className={styles.validationText}
            id={id && `${id}__validationText`}
          >
            {validationText}
          </div>
        )}
      </div>
    </label>
  );
};

TextField.defaultProps = {
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helpText: null,
  id: undefined,
  inputSize: null,
  isLabelVisible: true,
  layout: 'vertical',
  placeholder: null,
  required: false,
  size: 'medium',
  type: 'text',
  validationState: null,
  validationText: null,
  value: undefined,
  variant: 'outline',
};

TextField.propTypes = {
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line react/forbid-prop-types
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth: PropTypes.bool,
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element. It also serves as a prefix for nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id: PropTypes.string,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inputSize: PropTypes.number,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Text field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * Optional example value.
   */
  placeholder: PropTypes.string,
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Size of the field.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  /**
   * HTML input type, translated as `type` attribute of the input.
   */
  type: PropTypes.oneOf(['email', 'number', 'password', 'tel', 'text']),
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Design variant of the field, further customizable with CSS custom properties.
   */
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export const TextFieldWithContext = withForwardedRef(withProviderContext(TextField, 'TextField'));

export default TextFieldWithContext;
