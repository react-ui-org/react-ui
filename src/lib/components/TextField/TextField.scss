@use "../../styles/tools/form-fields/box-field-elements";
@use "../../styles/tools/form-fields/box-field-layout";
@use "../../styles/tools/form-fields/box-field-sizes";
@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";

// Foundation
.root {
    @include foundation.root();
}

.label {
    @include foundation.label();
}

.inputContainer {
    @include box-field-elements.input-container();
}

.input {
    @include box-field-elements.input();
}

.bottomLine {
    @include box-field-elements.bottom-line();
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.hasRootCustomInputSize .input {
    @include box-field-elements.input-size();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// Variants
.rootVariantFilled {
    @include variants.visual(box, $variant: filled);
}

.rootVariantOutline {
    @include variants.visual(box, $variant: outline);
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.rootLayoutVertical,
.rootLayoutHorizontal {
    @include box-field-layout.vertical();
}

.rootLayoutHorizontal {
    @include box-field-layout.horizontal();
}

.isRootFullWidth {
    @include box-field-layout.full-width();
}

.isRootInFormLayout {
    @include box-field-layout.in-form-layout();
}

.hasRootSmallInput.rootLayoutHorizontal {
    @include box-field-layout.horizontal-with-small-input();
}

// Sizes
.rootSizeSmall {
    @include box-field-sizes.size(small);
}

.rootSizeMedium {
    @include box-field-sizes.size(medium);
}

.rootSizeLarge {
    @include box-field-sizes.size(large);
}
