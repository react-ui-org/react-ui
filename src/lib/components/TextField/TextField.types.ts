export interface TextFieldProps {
  /**
   * Rest of the props.
   */
  [key: string]: unknown;
  /**
   * If `true`, the input will be disabled.
   */
  disabled?: boolean;
  /**
   * If `true`, the field will span the full width of its parent.
   */
  fullWidth?: boolean;
  /**
   * Optional help text.
   */
  helpText?: React.ReactNode;
  /**
   * ID of the input HTML element. It also serves as a prefix for nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id?: string;
  /**
   * Width of the input field. Translated as `size` attribute for input types other than `number`.
   */
  inputSize?: number;
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible?: boolean;
  /**
   * Text field label.
   */
  label: string;
  /**
   * Layout of the field.
   *
   * Ignored if the component is rendered within `FormLayout` component
   * as the value is inherited in such case.
   */
  layout?: Layout;
  /**
   * If `true`, the input will be required.
   */
  required?: boolean;
  /**
   * Size of the field.
   */
  size?: Size;
  /**
   * HTML input type, translated as `type` attribute of the input.
   */
  type?: InputType;
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState?: Validation;
  /**
   * Validation message to be displayed.
   */
  validationText?: React.ReactNode;
  /**
   * Design variant of the field, further customizable with CSS custom properties.
   */
  variant?: Exclude<Priority, 'flat'>;
}
