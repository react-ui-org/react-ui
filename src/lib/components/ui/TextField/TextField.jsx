import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './TextField.scss';

export const TextField = (props) => {
  const SMALL_INPUT_SIZE = 10;

  const setCustomInputSizeByType = (inputType, inputSize, max) => {
    if (inputType === 'number' && max) {
      return max.toString().length;
    }

    if (inputSize) {
      return inputSize;
    }

    return null;
  };

  const setInlineStyle = (customSize) => {
    const style = {};

    if (customSize) {
      style['--rui-custom-input-size'] = customSize;
    }

    return style;
  };

  const customInputSize = setCustomInputSizeByType(props.type, props.inputSize, props.max);
  const hasSmallInput = (customInputSize !== null) && (customInputSize <= SMALL_INPUT_SIZE);

  let labelVisibilityClass = '';
  let rootFullWidthClass = '';
  let rootInFormLayoutClass = '';
  let rootLayoutClass = '';
  let rootRequiredClass = '';
  let rootSizeClass = '';
  let rootSmallInputClass = '';
  let rootValidationStateClass = '';
  let rootVariantClass = '';

  if (!props.isLabelVisible) {
    labelVisibilityClass = styles.isLabelHidden;
  }

  if (props.fullWidth) {
    rootFullWidthClass = styles.isRootFullWidth;
  }

  if (props.inFormLayout) {
    rootInFormLayoutClass = styles.isRootInFormLayout;
  }

  if (hasSmallInput) {
    rootSmallInputClass = styles.hasRootSmallInput;
  }

  if (props.layout === 'horizontal') {
    rootLayoutClass = styles.rootLayoutHorizontal;
  } else if (props.layout === 'vertical') {
    rootLayoutClass = styles.rootLayoutVertical;
  }

  if (props.required) {
    rootRequiredClass = styles.isRootRequired;
  }

  if (props.size === 'small') {
    rootSizeClass = styles.rootSizeSmall;
  } else if (props.size === 'medium') {
    rootSizeClass = styles.rootSizeMedium;
  } else if (props.size === 'large') {
    rootSizeClass = styles.rootSizeLarge;
  }

  if (props.validationState === 'invalid') {
    rootValidationStateClass = styles.isRootStateInvalid;
  } else if (props.validationState === 'valid') {
    rootValidationStateClass = styles.isRootStateValid;
  } else if (props.validationState === 'warning') {
    rootValidationStateClass = styles.isRootStateWarning;
  }

  if (props.variant === 'filled') {
    rootVariantClass = styles.rootVariantFilled;
  } else if (props.variant === 'outline') {
    rootVariantClass = styles.rootVariantOutline;
  }

  const propsToTransfer = transferProps(
    props,
    ['changeHandler', 'disabled', 'fullWidth', 'helperText', 'id', 'inFormLayout', 'inputSize', 'isLabelVisible',
      'label', 'layout', 'max', 'placeholder', 'required', 'size', 'type', 'validationState', 'value', 'variant'],
  );

  return (
    <label
      className={(`
        ${styles.root}
        ${rootFullWidthClass}
        ${rootInFormLayoutClass}
        ${rootLayoutClass}
        ${rootRequiredClass}
        ${rootSizeClass}
        ${rootSmallInputClass}
        ${rootValidationStateClass}
        ${rootVariantClass}
      `).trim()}
      htmlFor={props.id}
      id={`${props.id}__label`}
      style={setInlineStyle(customInputSize)}
    >
      <div
        className={(`
          ${styles.label}
          ${labelVisibilityClass}
        `).trim()}
        id={`${props.id}__labelText`}
      >
        {props.label}
      </div>
      <div className={styles.field}>
        <div className={styles.inputContainer}>
          <input
            {...propsToTransfer}
            className={styles.input}
            disabled={props.disabled}
            id={props.id}
            max={props.type === 'number' ? props.max : null}
            onChange={props.changeHandler}
            placeholder={props.placeholder}
            ref={props.forwardedRef}
            required={props.required}
            size={props.type !== 'number' ? props.inputSize : null}
            type={props.type}
            value={props.value}
          />
          {props.variant === 'filled' && (
            <div className={styles.bottomLine} />
          )}
        </div>
        {props.helperText && (
          <div
            className={styles.helperText}
            id={`${props.id}__helperText`}
          >
            {props.helperText}
          </div>
        )}
      </div>
    </label>
  );
};

TextField.defaultProps = {
  changeHandler: null,
  disabled: false,
  forwardedRef: undefined,
  fullWidth: false,
  helperText: null,
  inFormLayout: false,
  inputSize: null,
  isLabelVisible: true,
  layout: 'vertical',
  max: null,
  placeholder: null,
  required: false,
  size: 'medium',
  type: 'text',
  validationState: null,
  value: undefined,
  variant: 'outline',
};

TextField.propTypes = {
  changeHandler: PropTypes.func,
  disabled: PropTypes.bool,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  fullWidth: PropTypes.bool,
  helperText: PropTypes.string,
  id: PropTypes.string.isRequired,
  inFormLayout: PropTypes.bool,
  inputSize: PropTypes.number,
  isLabelVisible: PropTypes.bool,
  label: PropTypes.string.isRequired,
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  max: PropTypes.number,
  placeholder: PropTypes.string,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  type: PropTypes.oneOf(['email', 'number', 'password', 'tel', 'text']),
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  variant: PropTypes.oneOf(['filled', 'outline']),
};

export default withForwardedRef(TextField);
