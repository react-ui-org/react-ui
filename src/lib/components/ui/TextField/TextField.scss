@import '../../../styles/tools/forms/foundation';
@import '../../../styles/tools/forms/layouts';
@import '../../../styles/tools/forms/relationships';
@import '../../../styles/tools/forms/sizes';
@import '../../../styles/tools/forms/states';
@import '../../../styles/tools/forms/variants';
@import '../../../styles/tools/accessibility';

// Foundation
.root {
  @include form-field-horizontal-neighbor();
}

.inputContainer {
  @include form-field-input-container();
}

.input {
  @include form-field-input();
  @include form-field-input-text();
}

.bottomLine {
  @include form-field-bottom-line();
}

.helperText {
  @include form-field-helper-text();
}

// Variants
.rootVariantFilled .input {
  @include form-field-input-variant(filled);
}

.rootVariantOutline .input {
  @include form-field-input-variant(outline);
}

// States
.isRootRequired .label {
  @include form-field-label-required();
}

.isRootStateInvalid {
  @include form-field-state(invalid);
}

.isRootStateValid {
  @include form-field-state(valid);
}

.isRootStateWarning {
  @include form-field-state(warning);
}

// Invisible label
.isLabelHidden {
  @include accessibility-hide-text();
}

// Layouts
.rootLayoutHorizontal {
  @include form-field-layout-horizontal();
}

.rootLayoutVertical {
  @include form-field-layout-vertical();
}

.isRootFullWidth {
  @include form-field-full-width();
  @include form-field-vertical-neighbor();
}

.hasRootSmallInput {
  @include form-field-layout-with-small-input();
}

// Sizes
.rootSizeSmall {
  @include form-field-size(small);
}

.rootSizeMedium {
  @include form-field-size(medium);
}

.rootSizeLarge {
  @include form-field-size(large);
}
