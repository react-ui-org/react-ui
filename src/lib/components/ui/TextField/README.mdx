---
name: TextField
menu: 'Components: UI'
route: /components/ui/text-field
---

# TextField

TextField allows users to input text information.

import {
  Playground,
  Props,
} from 'docz'
import { TextField } from './TextField'

## Basic Usage

To implement the TextField component, you need to import it first:

```js
import { TextField } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <TextField id="my-text-field" label="First name" />
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use the **most suitable input type** for current context: aside from the
  common `text` type, there are also `email`, `number`, `password`, and `tel`
  types at your disposal. A properly chosen input type is especially important
  for touch users as it triggers an appropriate virtual keyboard, so it helps
  speed up the completion of the field.

- **Beware of the `number` input type:** it may not be always what you want.
  Not all number-like values are actually numbers, eg. phone numbers, credit
  card numbers, or business IDs. In such cases use the most appropriate input
  type (probably `text` or `tel`) along with the
  [`pattern` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/pattern)
  to improve the input experience for touch users.

- Use **short and descriptive labels**, ideally nouns rather than seemingly
  polite phrases like _Please enter your first name_. Short labels will help
  your users accomplish their task faster.

- Only make the TextField's label invisible when there is **another visual
  clue** to guide users through filling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Optionally you can **pre-fill your text fields with a placeholder** to give
  users an example of what they're expected to fill in. Just remember the
  placeholder disappears once a value is entered so don't put anything important
  there. Usually, en example value starting with ‚ÄúEg.‚Äù works best.

- Use **clear, calm error messages** when there's a problem with what they
  entered. Be positive and focus on solutions to make the error message helpful.

- When asking users for their contact information or other personal information,
  make use of the
  [`autocomplete` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete)
  to help them fill the form faster.

## Design Variants

To satisfy the design requirements of your project, all input fields in React UI
come in **two design variants** to choose from: outline and filled. Both can be
further [customized](/customize/theming) with CSS custom properties.

<Playground>
  <TextField
    id="my-text-field-outline"
    label="First name"
  />
  <TextField
    id="my-text-field-filled"
    label="First name"
    variant="filled"
  />
</Playground>

## Sizes

Aside from the default (medium) size, two additional sizes are available: small
and large.

<Playground>
  <TextField
    id="my-text-field-outline-small"
    label="First name"
    size="small"
  />
  <TextField
    id="my-text-field-outline-medium"
    label="First name"
  />
  <TextField
    id="my-text-field-outline-large"
    label="First name"
    size="large"
  />
  <TextField
    id="my-text-field-filled-small"
    label="First name"
    size="small"
    variant="filled"
  />
  <TextField
    id="my-text-field-filled-medium"
    label="First name"
    variant="filled"
  />
  <TextField
    id="my-text-field-filled-large"
    label="First name"
    size="large"
    variant="filled"
  />
</Playground>

Full-width fields span the full width of a parent:

<Playground>
  <TextField
    id="my-text-field-outline-full-width"
    label="First name"
    fullWidth
  />
  <TextField
    id="my-text-field-filled-full-width"
    label="First name"
    variant="filled"
    fullWidth
  />
</Playground>

## Input Types

Aside from the common `text` type, there are also `email`, `number`, `password`,
and `tel` types at your disposal.

<Playground>
  <TextField
    id="my-text-field-outline-input-type-email"
    label="Email"
    type="email"
  />
  <TextField
    id="my-text-field-filled-input-type-email"
    label="Email"
    variant="filled"
    type="email"
  />
  <TextField
    id="my-text-field-outline-input-type-number"
    label="Age"
    type="number"
  />
  <TextField
    id="my-text-field-filled-input-type-number"
    label="Age"
    variant="filled"
    type="number"
  />
  <TextField
    id="my-text-field-outline-input-type-password"
    label="Password"
    type="password"
  />
  <TextField
    id="my-text-field-filled-input-type-password"
    label="Password"
    variant="filled"
    type="password"
  />
  <TextField
    id="my-text-field-outline-input-type-tel"
    label="Phone"
    type="tel"
  />
  <TextField
    id="my-text-field-filled-input-type-tel"
    label="Phone"
    variant="filled"
    type="tel"
  />
</Playground>

## Input Size

The default width of all inputs is 240 px, and it can be
[customized](/customize/theming) with a CSS custom property. However, you
can also **change the width of individual text fields** using the `inputSize`
property. It (obviously) sets the `size` attribute of the `input` element and is
further picked up by CSS to normalize rendering across browsers.

üëâ Remember that the
[`size`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/text#size)
HTML attribute **does not limit on how many characters** the user can enter. Use
the `maxlength` attribute to achieve that effect.

üëâ For compatibility with the HTML specification, the `inputSize` option (which
is translated as `size` attribute of the input element) is
[not available for the `number` input type](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#Controlling_input_size).
However, you can use the `max` attribute to determine the desired width of the
number input instead.

<Playground>
  <TextField
    id="my-text-field-outline-input-size"
    inputSize={3}
    label="Title"
  />
  <TextField
    id="my-text-field-filled-input-size"
    inputSize={3}
    label="Title"
    variant="filled"
  />
  <TextField
    id="my-text-field-outline-input-size-number"
    label="Age"
    min={13}
    max={120}
    type="number"
  />
  <TextField
    id="my-text-field-filled-input-size-number"
    label="Age"
    min={13}
    max={120}
    variant="filled"
    type="number"
  />
</Playground>

## Forwarding HTML Attributes

When you want to improve the accessibility of your text fields even further, you
can **add whatever HTML attribute you like.** All attributes that don't
interfere with [component's API](#api) are forwarded to the native HTML input.

<Playground>
  <TextField
    id="my-text-field-outline-forwarding"
    label="Address"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
  <TextField
    id="my-text-field-filled-forwarding"
    label="Address"
    variant="filled"
    autoComplete="street-address"
    minLength={3}
    maxLength={80}
  />
</Playground>

üëâ Refer to the MDN reference for the full list of supported attributes of the
[text](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/text),
[email](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/email),
[number](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number),
[tel](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/tel), and
[password](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/password)
input types.

## Invisible Label

In some cases, it may be convenient to visually hide the field label. The label
remains accessible to assistive technologies.

While it may be acceptable for login screens with just a few fields or for other
simple forms, it's dangerous to hide labels from users in most cases. Keep in
mind you should **provide another visual clue** so users know what to fill into
the input.

<Playground>
  <TextField
    id="my-text-field-outline-hidden-label"
    label="User name"
    isLabelVisible={false}
    placeholder="Eg. john@example.com"
  />
  <TextField
    id="my-text-field-filled-hidden-label"
    label="User name"
    isLabelVisible={false}
    variant="filled"
    placeholder="Eg. john@example.com"
  />
</Playground>

## Horizontal Layout

The default vertical layout is very easy to use and work with. However, there
are situations where horizontal layout suits better ‚Äî and that's why React UI
supports this kind of layout as well.

<Playground>
  <TextField
    id="my-text-field-outline-input-size-horizontal"
    inputSize={3}
    label="Title"
    layout="horizontal"
  />
  <TextField
    id="my-text-field-filled-input-size-horizontal"
    inputSize={3}
    label="Title"
    layout="horizontal"
    variant="filled"
  />
  <TextField
    id="my-text-field-outline-horizontal"
    label="First name"
    layout="horizontal"
  />
  <TextField
    id="my-text-field-filled-horizontal"
    label="First name"
    layout="horizontal"
    variant="filled"
  />
  <TextField
    id="my-text-field-outline-horizontal-full-width"
    label="First name"
    layout="horizontal"
    fullWidth
  />
  <TextField
    id="my-text-field-filled-horizontal-full-width"
    label="First name"
    layout="horizontal"
    variant="filled"
    fullWidth
  />
  <TextField
    id="my-text-field-outline-horizontal-full-width-invisible-label"
    isLabelVisible={false}
    label="First name"
    layout="horizontal"
    placeholder="First name"
    fullWidth
  />
  <TextField
    id="my-text-field-filled-horizontal-full-width-invisible-label"
    isLabelVisible={false}
    label="First name"
    layout="horizontal"
    placeholder="First name"
    variant="filled"
    fullWidth
  />
</Playground>

## Help Text

You may provide an additional help text to clarify how the input should be
filled.

<Playground>
  <TextField
    id="my-text-field-outline-help-text"
    label="First name"
    helperText="Choose one or more kinds of fruit to feel happy."
  />
  <TextField
    id="my-text-field-filled-help-text"
    label="First name"
    variant="filled"
    helperText="Choose one or more kinds of fruit to feel happy."
  />
  <TextField
    id="my-text-field-outline-help-text-horizontal"
    label="First name"
    layout="horizontal"
    helperText="Choose one or more kinds of fruit to feel happy."
  />
  <TextField
    id="my-text-field-filled-help-text-horizontal"
    label="First name"
    layout="horizontal"
    helperText="Choose one or more kinds of fruit to feel happy."
    variant="filled"
  />
  <TextField
    id="my-text-field-outline-help-text-horizontal-full-width"
    label="First name"
    layout="horizontal"
    helperText="Choose one or more kinds of fruit to feel happy."
    fullWidth
  />
  <TextField
    id="my-text-field-filled-help-text-horizontal-full-width"
    label="First name"
    layout="horizontal"
    variant="filled"
    helperText="Choose one or more kinds of fruit to feel happy."
    fullWidth
  />
</Playground>

Keep in mind that **long help texts don't play well with small input sizes,**
especially in a vertical layout. To fix this at least for the horizontal layout,
the help text expands over the full field width when the desired input width
(based on `inputSize` and `max` options) is 10 characters or smaller.

<Playground>
  <TextField
    id="my-text-field-outline-help-text-input-size-horizontal"
    inputSize={3}
    label="Title"
    layout="horizontal"
    helperText="What academic degree do you hold?"
  />
  <TextField
    id="my-text-field-filled-help-text-input-size-horizontal"
    inputSize={3}
    label="Title"
    layout="horizontal"
    variant="filled"
    helperText="What academic degree do you hold?"
  />
  <TextField
    id="my-text-field-outline-help-text-input-size-number-horizontal"
    label="Age"
    layout="horizontal"
    min={13}
    max={120}
    type="number"
    helperText="How old do you see yourself?"
  />
  <TextField
    id="my-text-field-filled-help-text-input-size-number-horizontal"
    label="Age"
    layout="horizontal"
    min={13}
    max={120}
    variant="filled"
    type="number"
    helperText="How old do you see yourself?"
  />
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide a help text for states other than valid** so users know
what happened and what action they should take or what options they have.

<Playground>
  <TextField
    id="my-text-field-outline-success"
    label="User name"
    validationState="valid"
    helperText="Looks good!"
    changeHandler={() => {}}
    value="johnny1986"
    required
  />
  <TextField
    id="my-text-field-outline-warning"
    label="User name"
    validationState="warning"
    helperText="Account with this name already exists, pick a different one."
    value="joe"
    changeHandler={() => {}}
    required
  />
  <TextField
    id="my-text-field-outline-error"
    label="User name"
    validationState="invalid"
    helperText="Please enter a user name of your choice."
    required
  />
  <TextField
    id="my-text-field-filled-success"
    label="User name"
    validationState="valid"
    variant="filled"
    helperText="Looks good!"
    changeHandler={() => {}}
    value="johnny1986"
    required
  />
  <TextField
    id="my-text-field-filled-warning"
    label="User name"
    validationState="warning"
    variant="filled"
    helperText="Account with this name already exists, pick a different one."
    changeHandler={() => {}}
    value="joe"
    required
  />
  <TextField
    id="my-text-field-filled-error"
    label="User name"
    validationState="invalid"
    variant="filled"
    helperText="Please enter a user name of your choice."
    required
  />
</Playground>

### Disabled State

It's possible to disable the whole input.

<Playground>
  <TextField
    id="my-text-field-outline-disabled"
    label="First name"
    value="John"
    disabled
  />
  <TextField
    id="my-text-field-filled-disabled"
    label="First name"
    variant="filled"
    value="John"
    disabled
  />
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `input` HTML element.
This is useful mainly to improve component's
[accessibility](#forwarding-html-attributes).

<Props of={TextField} />

---

Next: [Table ‚Üí](/components/ui/table)
