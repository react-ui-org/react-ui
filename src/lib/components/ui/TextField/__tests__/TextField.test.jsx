import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { TextField } from '../TextField';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<TextField id="test" label="label" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with hidden label', () => {
    const tree = shallow(<TextField
      id="test"
      isLabelVisible={false}
      label="With hidden label"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with small input', () => {
    const tree = shallow(<TextField
      id="test"
      inputSize={5}
      label="Small input"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly the number input type', () => {
    const tree = shallow(<TextField
      id="test"
      label="Number input type"
      type="number"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly the number input type with small input', () => {
    const tree = shallow(<TextField
      id="test"
      label="Number input type"
      inputSize={1}
      type="number"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with custom props', () => {
    const tree = shallow(<TextField
      autoCapitalize="off"
      autoComplete="off"
      id="test"
      label="With custom props"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<TextField
      disabled
      fullWidth
      helpText="some help"
      id="test"
      inFormLayout
      inputSize={20}
      isLabelVisible={false}
      label="All props"
      layout="horizontal"
      placeholder="placeholder"
      required
      size="large"
      type="email"
      validationState="invalid"
      validationText="some error"
      value="value"
      variant="filled"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls changeHandler()', () => {
    const spy = sinon.spy();
    const component = mount(<TextField
      changeHandler={spy}
      id="test"
      label="label"
    />);

    component
      .find('input')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'test' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
