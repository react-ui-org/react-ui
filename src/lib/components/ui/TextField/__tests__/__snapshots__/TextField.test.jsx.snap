// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
  style={Object {}}
>
  <div
    className="label"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      disabled={false}
      id="test"
      onChange={null}
      placeholder={null}
      required={false}
      size={null}
      type="text"
    />
  </div>
</label>
`;

exports[`rendering renders correctly with all props 1`] = `
<label
  className="root
        isRootFullWidth
        isRootInFormLayout
        rootLayoutHorizontal
        isRootRequired
        rootSizeLarge
        hasRootSmallInput
        isRootStateInvalid
        rootVariantFilled"
  htmlFor="test"
  id="test__label"
  style={
    Object {
      "--rui-custom-input-size": 3,
    }
  }
>
  <div
    className="label
          isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      disabled={true}
      id="test"
      onChange={null}
      placeholder="placeholder"
      required={true}
      size={3}
      type="email"
      value="value"
    />
    <div
      className="bottomLine"
    />
  </div>
  <div
    className="helperText"
    id="test__helperText"
  >
    some help
  </div>
</label>
`;

exports[`rendering renders correctly with hidden label 1`] = `
<label
  className="root
        
        
        rootLayoutVertical
        
        rootSizeMedium
        
        
        rootVariantOutline"
  htmlFor="test"
  id="test__label"
  style={Object {}}
>
  <div
    className="label
          isLabelHidden"
    id="test__labelText"
  >
    label
  </div>
  <div
    className="inputContainer"
  >
    <input
      className="input"
      disabled={false}
      id="test"
      onChange={null}
      placeholder={null}
      required={false}
      size={null}
      type="text"
    />
  </div>
</label>
`;
